[tests] Speed up fuzzing by ~200x when using afl-fuzz

practicalswift opened the pull request:
Enable the `afl-clang-fast++` features deferred forkserver (`__AFL_INIT`) and persistent mode (`__AFL_LOOP(1000)`).

Before this patch:

```
$ afl-fuzz -i input -o output -m512 -- src/test/test_bitcoin_fuzzy
[*] Validating target binary...
[!] WARNING: The target binary is pretty slow! See /usr/local/share/doc/afl/perf_tips.txt.
[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 1 total
       Bitmap range : 1072 to 1072 bits (average: 1072.00 bits)
        Exec timing : 20.4k to 20.4k us (average: 20.4k us)
…
exec speed : 57.58/sec (slow!)
exec speed : 48.35/sec (slow!)
exec speed : 53.78/sec (slow!)
```

After this patch:

```
$ afl-fuzz -i input -o output -m512 -- src/test/test_bitcoin_fuzzy
[*] Validating target binary...
[+] Persistent mode binary detected.
[+] Deferred forkserver binary detected.
[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 1 total
       Bitmap range : 24 to 24 bits (average: 24.00 bits)
        Exec timing : 114 to 114 us (average: 114 us)
…
exec speed : 15.9k/sec
exec speed : 13.1k/sec
exec speed : 15.1k/sec
```

For more fuzzing discussions, see #10364.

laanwj commented:
Interesting. Concept ACK.

fanquake commented:
I'm currently seeing no speedup when using this new code. I'm wondering if it's because I have to run with AFL_NO_FORKSRV on OS X (I haven't been able to make afl-fuzz run without it being set) ?

Or does this require using [afl-clang-fast and afl-clang-fast++](https://github.com/mirrorer/afl/tree/master/llvm_mode) in place of afl-clang and afl-clang++ to see the speed improvements?

I'm seeing:
```
[*] Attempting dry run with 'id:000566,orig:id:000958,sync:39,src:000955'...
    len = 7368, map size = 1430, exec speed = 15345 us
[+] All test cases processed.

[!] WARNING: The target binary is pretty slow! See /usr/local/Cellar/afl-fuzz/2.41b/share/doc/afl/perf_tips.txt.
[+] Here are some useful stats:

    Test case count : 141 favored, 4 variable, 567 total
       Bitmap range : 995 to 1684 bits (average: 1240.06 bits)
        Exec timing : 13.9k to 90.9k us (average: 15.6k us)

[*] No -t option specified, so I'll use exec timeout of 100 ms.
[+] All set and ready to roll!
```

practicalswift commented:
@fanquake The binaries you are running are not instrumented correctly (the messages `Persistent mode binary detected` and `Deferred forkserver binary detected` are missing). You're correct that this speedup requires using `afl-clang-fast` and `afl-clang-fast++` in place of `afl-clang` and `afl-clang++`. See https://github.com/mirrorer/afl/tree/master/llvm_mode for details :-)

practicalswift committed:
[test] Speed up fuzzing by ~200x when using afl-fuzz

Enable the `afl-clang-fast++` features deferred forkserver (`__AFL_INIT`) and persistent mode (`__AFL_LOOP(1000)`).

Before this patch:

```
$ afl-fuzz -i input -o output -m512 -- src/test/test_bitcoin_fuzzy
[*] Validating target binary...
[!] WARNING: The target binary is pretty slow! See /usr/local/share/doc/afl/perf_tips.txt.
[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 1 total
       Bitmap range : 1072 to 1072 bits (average: 1072.00 bits)
        Exec timing : 20.4k to 20.4k us (average: 20.4k us)
…
exec speed : 57.58/sec (slow!)
exec speed : 48.35/sec (slow!)
exec speed : 53.78/sec (slow!)
```

After this patch:

```
$ afl-fuzz -i input -o output -m512 -- src/test/test_bitcoin_fuzzy
[*] Validating target binary...
[+] Persistent mode binary detected.
[+] Deferred forkserver binary detected.
[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 1 total
       Bitmap range : 24 to 24 bits (average: 24.00 bits)
        Exec timing : 114 to 114 us (average: 114 us)
…
exec speed : 15.9k/sec
exec speed : 13.1k/sec
exec speed : 15.1k/sec
```

fanquake commented:
In that case, the code changes should be accompanied by a new? section in the fuzzing documentation, outlining how to setup and run the much faster fuzzing.

practicalswift commented:
@fanquake Good point! Documentation and comments added. Looks good? :-)

practicalswift commented:
Related pull request: #10440 ("Add libFuzzer support").

laanwj commented:
@fanquake thanks for testing
@practicalswift thanks for adding documentation
utACK https://github.com/bitcoin/bitcoin/pull/10415/commits/693247b82bbccada37f7bc61380ddac799100eb8
